### DISCUSSION TITLE
Comparison with py-spy?

### DISCUSSION THREAD

**Discussion by nalzok** [bloomberg/pystack#103] [General]
I wonder how does this project compare to [py-spy](https://github.com/benfred/py-spy)?

**Comment by pablogsal** [comment]
The main difference is that `py-spy` is a profiler while `pystack` is a debugger. This means that while both share functionality, `pystack` focus is getting as much information as possible even in the most complicated situations at the expense of time. This means that `pystack` works in some situations where `py-spy` does not. Currently, `pystack` is the only one to support core files as well (although IIRC they have plans to support some core file formats in the future). Here is a short table comparing what I know are the difference between both tools:

| Feature | py-spy | pystack |
|---------|--------|---------|
| Works on Linux | ✔️ | ✔️ |
| Works on MacOS | ✔️ | ❌ |
| Works on Windows | ✔️ | ❌ |
| Performance profiler (focus on speed) | ✔️ | ❌ |
| Debugger (focus on information and correctness) | ❌ | ✔️ |
| Can profile a process to tell why is slow | ✔️ | ❌ |
| Has an htop like interface | ✔️ | ❌ |
| Works with running processes | ✔️ | ✔️ |
| Works with core dump files | ❌ | ✔️ |
| Shows Python GIL status | ✔️ | ✔️ |
| Shows garbage collection cycles | ✔️ | ✔️ |
| Shows native function calls | ✔️ | ✔️ |
| Demangles symbols in native stack | ✔️ | ✔️ |
| Includes inlined functions in native stack | ❌ | ✔️ |
| Shows values of local variables and function arguments | ✔️ | ✔️ |
| Safe to use on running processes | ✔️ | ✔️ |
| Perform Python stack analysis without pausing process | ✔️ | ✔️ |
| Works with optimized Python interpreter binaries | ❓ | ✔️ |
| Works without symbols or debug information | ❓ | ✔️ |
| Tolerates memory corruption | ❌ | ✔️ |
| Uses PEP 657 information in tracebacks | ❌ | ✔️ |
